### ISSUE TITLE
Tile loading goes into infinite loop

### ISSUE THREAD

**Issue by pmatos** [alex-hhh/ActivityLog2#29]
Map tile loading after loading as finished goes into an infinite loop and seems to keep clearing and loading new tiles.

https://www.dropbox.com/s/z28idqwfnom08tx/simplescreenrecorder-2018-07-02_09.42.06.ogg?dl=0

(video is in my dropbox as it seems github doesn't allow direct attachments of videos here)

**Comment by alex-hhh** [comment]
This looks like a bug.  Before we try anything else, can you change the redraw interval from 100 to 2000 in the line below and see if things improve?

https://github.com/alex-hhh/ActivityLog2/blob/master/rkt/widgets/map-widget/map-widget.rkt#L580

Also, are you running the latest git revision or the 1.5.0 release?

Thanks,
Alex.

**Comment by pmatos** [comment]
Interesting. This doesn't happen on my laptop. Need to check it out once I get home.

**Comment by pmatos** [comment]
I guess you meant line 595 above:`           (send redraw-timer start 100))))`
https://github.com/alex-hhh/ActivityLog2/blob/27f56adc65dda7df21997c67144c9f8048bcc510/rkt/widgets/map-widget/map-widget.rkt#L595

**Comment by pmatos** [comment]
Same problem with `2000`. The tile loop is slightly slower but I can still see the same issue I showed above. This is a 4K screen so it might have to do with it. Laptop doesn't have this issue.

**Comment by alex-hhh** [comment]
I also have a 4k display on my laptop and the code works fine.  I tried it both with 1.0 scaling and 1.5 scaling(which I use by default).  The map renders fine on my machine, so it is not the 4k display that is the problem.

Unfortunately I cannot reproduce the problem, so  I will have to ask you to do some diagnostics investigations, if this is OK with you...

----

Can you resize the application window so that it is smaller, and see if the problem still happens?

----

The code draws a white bitmap when it has no tile, but this is indistinguishable from the map background

https://github.com/alex-hhh/ActivityLog2/blob/27f56adc65dda7df21997c67144c9f8048bcc510/rkt/widgets/map-widget/map-widget.rkt#L553

Can you add the following code after the line above:

```racket
(let ((dc (new bitmap-dc% [bitmap empty-bmp])))
      (define brush (send the-brush-list find-or-create-brush "red" 'cross-hatch))
      (send dc set-brush brush)
      (send dc draw-rectangle 0 0 tile-size tile-size))
```

This will make the empty tile a red crosshatch tile, so it will be distinguishable from an empty canvas.

----

The canvas is painted in the `on-canvas-paint` method and this method is only invoked as a response to GUI refresh events (sometimes the code requests a refresh, but it never invokes the method directly:

https://github.com/alex-hhh/ActivityLog2/blob/27f56adc65dda7df21997c67144c9f8048bcc510/rkt/widgets/map-widget/map-widget.rkt#L506

Here are a few things to try:

* first, add a `printf` message to this method, to print out that it has been called (this is the easiest way to visualize a redraw)
* comment out the `(draw-map-tiles canvas dc)` line.  The map tiles will not be drawn, but the GPS track will.  You can check if the view still flickers and also if the `on-canvas-paint` is actually called the you don't move the map (it should not be called in this case, as a redraw is only requested when tiles are downloaded or when you pan the map).

if flickering stops, the problem is in the `draw-map-tiles`, if it does not, there is a problem with the canvas refresh.  I have a few more things to try, but I would like first to establish if `draw-map-tiles` is OK or not.

**Comment by alex-hhh** [comment]
Hi @pmatos, did you have a chance to try out the suggestions I made? Thanks, Alex.

**Comment by pmatos** [comment]
Sorry for this. I have been having so much work lately I haven't managed to try it yet. One of the things I need to sort out to be able to use AL2 on a regular basis is the automation with the activities imports. With 2-3 sessions per day, doing it manually is quite time consuming in the long run.

**Comment by pmatos** [comment]
On this topic, I am trying to integrate https://github.com/pe-st/garmin-connect-export/tree/develop into my workflow reading files overnight, and then calling al2 on the cli only to import the files without starting the gui. It seem to need some changes but it shouldn't be too hard to make it work.

**Comment by pmatos** [comment]
> Can you resize the application window so that it is smaller, and see if the problem still happens?

This only happens past a certain size. Take a look at the following screenshots. One is file, the other is a few pixels wider and shows the problem.
![2018-07-13-094352_2814x1644_scrot](https://user-images.githubusercontent.com/7911/42679203-686e66ee-8681-11e8-8d51-2bb1046ed145.png)
![2018-07-13-094412_2830x1638_scrot](https://user-images.githubusercontent.com/7911/42679204-68869d4a-8681-11e8-88f1-cb3a8c50de77.png)

This is around the 2830px of *map* width.

**Comment by pmatos** [comment]
> This will make the empty tile a red crosshatch tile, so it will be distinguishable from an empty canvas.

Shows the issue with the different background.
![2018-07-13-095250_3840x2160_scrot](https://user-images.githubusercontent.com/7911/42679570-b73b2ffe-8682-11e8-96e8-9caad2516675.png)

**Comment by pmatos** [comment]
`draw-map-tiles` looks good. The canvas is redrawn no matter what. The `on-canvas-paint` is constantly called, even when I don't do anything. Who calls this? Could it be a racket issue at this point? Still strange that we only see a loop past a certain window width. 

I also noticed that if I pan the map to certain places, the redrawing issue stops. If I pan it to other places, the loop restarts.

**Comment by pmatos** [comment]
I have now found, with a few extra printf that this when the loop is occuring, it always setting `request-redraw?` to `#t`.
https://github.com/alex-hhh/ActivityLog2/blob/a0ed9383d55a0463cecce589490ba3c2894db052/rkt/widgets/map-widget/map-widget.rkt#L593

**Comment by pmatos** [comment]
So I have added some printfs and see the following:
```
on-canvas-paint 1531470148139
drawing map tiles - display-backing-scale: 1.0 - 1531470148139
request-redraw? zoom-level 12 tile-x 2170 tile-y 1402
request-redraw? zoom-level 12 tile-x 2170 tile-y 1403
request-redraw? zoom-level 12 tile-x 2170 tile-y 1404
redraw-timer start: requ
est-redraw? #t, download-backlog 0, time 1531470148162
on-canvas-paint 1531470148262
drawing map tiles - display-backing-scale: 1.0 - 1531470148262
request-redraw? zoom-level 12 tile-x 2169 tile-y 1407
request-redraw? zoom-level 12 tile-x 2170 tile-y 1400
request-redraw? zoom-level 12 tile-x 2170 tile-y 1401
redraw-timer start: request-redraw? #t, download-backlog 0, time 1531470148288
on-canvas-paint 1531470148388
drawing map tiles - display-backing-scale: 1.0 - 1531470148388
request-redraw? zoom-level 12 tile-x 2169 tile-y 1404
request-redraw? zoom-level 12 tile-x 2169 tile-y 1405
request-redraw? zoom-level 12 tile-x 2169 tile-y 1406
redraw-timer start: request-redraw? #t, download-backlog 0, time 1531470148415
on-canvas-paint 1531470148515
drawing map tiles - display-backing-scale: 1.0 - 1531470148515
request-redraw? zoom-level 12 tile-x 2169 tile-y 1401
request-redraw? zoom-level 12 tile-x 2169 tile-y 1402
request-redraw? zoom-level 12 tile-x 2169 tile-y 1403
redraw-timer start: request-redraw? #t, download-backlog 0, time 1531470148538
on-canvas-paint 1531470148639
drawing map tiles - display-backing-scale: 1.0 - 1531470148639
request-redraw? zoom-level 12 tile-x 2168 tile-y 1406
request-redraw? zoom-level 12 tile-x 2168 tile-y 1407
request-redraw? zoom-level 12 tile-x 2169 tile-y 1400
redraw-timer start: request-redraw? #t, download-backlog 0, time 1531470148665
on-canvas-paint 1531470148766
drawing map tiles - display-backing-scale: 1.0 - 1531470148766
request-redraw? zoom-level 12 tile-x 2168 tile-y 1403
request-redraw? zoom-level 12 tile-x 2168 tile-y 1404
request-redraw? zoom-level 12 tile-x 2168 tile-y 1405
redraw-timer start: request-redraw? #t, download-backlog 0, time 1531470148792
```

Large number is current milliseconds. Not sure why `request-redraw?` is coming 3 at a time with close values for `tile-y` but it might be a clue to you.

**Comment by alex-hhh** [comment]
OK, I think I know what the problem is.   Can you change `cache-threshold` to 1000 in map-tiles.rkt and see if this fixes the problem? 1000 is probably too large, but I want to be sure the cache is not rotated too early.

https://github.com/alex-hhh/ActivityLog2/blob/a0ed9383d55a0463cecce589490ba3c2894db052/rkt/widgets/map-widget/map-tiles.rkt#L507

----

To make a CLI file importer, all you need to do is open the data base using this code:

```racket
(require "dbapp.rkt" "utilities.rkt")
(define *db*
  (let ((db-file (get-pref 'activity-log:database-file (lambda () #f))))
    (unless db-file (error "No default database, open one in the ActivityLog2 application."))
    (begin0
        (open-activity-log db-file)
      (printf "Opened ~a~%" db-file))))
(when *db* (set-current-database *db*))
```

Than call `import-new-activities-from-directory` from "import.rkt" with the database and the directory you want to import from.

If you create a CLI application for importing activities, I would be interested in adding it to ActivityLog2

----

For my part, I have a Raspberry PI that I use to download FIT files off my Garmin device, while charging the device.  I mounted the share with the FIT files on my laptop and run an "import from directory" inside ActivityLog2. Since it remembers the last import directory, this is just a "Ctrl-I + ENTER".  The project is described here:  https://github.com/alex-hhh/PiFitSync

**Comment by pmatos** [comment]
That's correct. Increasing the number of cached tiles fixes it. I can understand why as well. I guess that this happened because the number of tiles on the screen was larger than the number of cached tiles? And therefore it went into a loop to get more?

Probably the number of cached tiles should be at least as large as the number shown on screen?

**Comment by pmatos** [comment]
Thanks for the CLI importer snippet, I will integrate into what I have and create a pull request.

**Comment by alex-hhh** [comment]
I fixed this both on `master` and on `rel-1.5.0-fixes` branch.  I also added some logging to see how many tiles are set as a threshold, you can check that in `ActivityLogDbg.log` file.

The red tiles were really helpful, as it allowed me to differentiate between a white background of the canvas itself and the white background of the empty bitmap, which is drawn when a tile is not available immediately.  Since the background was red, I knew that the map widget was drawing empty bitmaps, from then on, it didn't take too much to figure it out...
